wuchong/flink
# address review comments, it mostly includes:

Thanks for contributing to Apache Flink. Before you open your pull request, please take the following check list into consideration.
If your changes take all of the items into account, feel free to open your pull request. For more information and/or questions please refer to the [How To Contribute guide](http://flink.apache.org/how-to-contribute.html).
In addition to going through the list, please provide a meaningful description of your changes.

- [ ] General
  - The pull request references the related JIRA issue ("[FLINK-XXX] Jira title text")
  - The pull request addresses only one issue
  - Each commit in the PR has a meaningful commit message (including the JIRA id)

  - Documentation has been added for new functionality
  - Old documentation affected by the pull request has been updated
  - JavaDoc for public methods has been added

- [ ] Tests & Build
  - Functionality added by the pull request is covered by tests
  - `mvn clean verify` has been executed successfully locally or a Travis build has passed

1. Forbid TableFunction implemented by Scala object, since the `collect` is called on a singleton. It will be error-prone in some concurrent cases.
2. Make `UserDefinedFunction` clean, and move eval relative functions and `createSqlFunction` to utils.
3. Rename `ScalarFunctions` to `SqlFunctions` because contains also TableFunction logic.
4. Restructure tests. Test Java Table API by comparing the RelNode of two tables. And check SQL API's DataSetRel and DataStreamRel via `TableTestBase` utils.
5. Scala Table API implicitly convert `TableFunction` into `TableFunctionCall`. `TableFunctionCall` is not an `Expression` or `LogicalNode`, but is like `GroupWindow`, can be visible to the users (contains API such `as(...)`).
6. Fix the hierarchy type extraction problem.
7. Check correct errors if a invalid table function is called or registered.
